### PR TITLE
Add wildcard for arrays

### PR DESCRIPTION
**Pull request by ostap0207**
INF-527

### CODE REVIEW

File: spec/preprocessors/whitelist_spec.rb [hunk 1]
```diff
@@ -74,6 +74,28 @@
     end
   end
 
+  context 'with whitelisted array elements field with wildcard' do
+    let(:data) {{
+      array: [{field: 'data', secret: 'secret'}]
```
Comment (ostankin) [inline spec/preprocessors/whitelist_spec.rb]: Not really critical, but I would make it `[{field: 'data1', secret: 'secret1'}, {field: 'data2', secret: 'secret2'}]` for peace of mind :)

File: spec/preprocessors/whitelist_spec.rb [hunk 2]
```diff
@@ -149,4 +171,15 @@
       expect(processed_data).to include('field_with_~'=> 'secret')
     end
   end
+  
+  context 'when field has tilde with 0 and 1' do
```
Comment (ostankin) [inline spec/preprocessors/whitelist_spec.rb]: The title says "tilde with 0 and 1", but the test contains only "tilde with 0". If I understand correctly, `~1` represents `/`, and there doesn't seem to be a test for this case.

**Comment by ostankin** [comment]
Minor comments regarding the tests, otherwise 👍

File: spec/preprocessors/whitelist_spec.rb [hunk 3]
```diff
@@ -149,4 +173,15 @@
       expect(processed_data).to include('field_with_~'=> 'secret')
     end
   end
+  
+  context 'when field has tilde and 1' do
+    let(:data) {{
+      'field_with_~1' => 'secret'
+    }}
```
Comment (ChameleonTartu) [inline spec/preprocessors/whitelist_spec.rb]: I think, it would be a bit better:
```
let(:data) { {'field_with_~1' => 'secret'} }
```

**Comment by ChameleonTartu** [comment]
General comment. Wouldn't be it better to include description as well, what is done, what is not ? Because without looking into the ticket is quite hard to understand what is going on why this or that code removed/present.

File: lib/logasm/preprocessors/whitelist.rb [hunk 4]
```diff
@@ -4,16 +4,25 @@ class Whitelist
...
       class InvalidPointerFormatException < Exception
       end
 
       def initialize(config = {})
         pointers = (config[:pointers] || []) + DEFAULT_WHITELIST
-        @fields_to_include = pointers.inject({}) do |mem, pointer|
+        decoded_pointers = pointers.inject([]) do |mem, pointer|
           validate_pointer(pointer)
-          mem.merge(decode(pointer) => true)
+          mem << decode(pointer)
```
Comment (indrekj) [inline lib/logasm/preprocessors/whitelist.rb]: `mem + [decode(pointer)]` to avoid mutation

Comment (indrekj) [inline lib/logasm/preprocessors/whitelist.rb]: hmm, actually would `map` work?
```
decoded_pointers = pointers.
  .each(&method(:validate_pointer))
  .map(&method(:decode))
```

File: lib/logasm/preprocessors/whitelist.rb [hunk 5]
```diff
@@ -22,6 +31,34 @@ def process(data)
 
       private
 
+
+      def has_wildcard(pointer)
```
Comment (indrekj) [inline lib/logasm/preprocessors/whitelist.rb]: `has_wildcard?`

File: lib/logasm/preprocessors/whitelist.rb (hunk 4 above)
Comment (ostap0207) [inline lib/logasm/preprocessors/whitelist.rb]: Yeah, a map is better.

File: README.md [hunk 6]
```diff
@@ -108,27 +108,28 @@ Received request {"info":{"phone":"************"}}
 
 Masks all the fields except those whitelisted in the configuration using [JSON Pointer](https://tools.ietf.org/html/rfc6901).
 Only simple values(`string`, `number`, `boolean`) can be whitelisted.
+Whitelisting array elements can be done using wildcard symbol `~` instead of element index.
```
Comment (indrekj) [inline README.md]: `instead of element index` part seems irrelevant.

File: lib/logasm/preprocessors/whitelist.rb [hunk 7]
```diff
@@ -4,16 +4,24 @@ class Whitelist
...
       class InvalidPointerFormatException < Exception
       end
 
       def initialize(config = {})
         pointers = (config[:pointers] || []) + DEFAULT_WHITELIST
-        @fields_to_include = pointers.inject({}) do |mem, pointer|
-          validate_pointer(pointer)
-          mem.merge(decode(pointer) => true)
+        decoded_pointers = pointers
+         .each(&method(:validate_pointer))
```
Comment (indrekj) [inline lib/logasm/preprocessors/whitelist.rb]: indendation error, missing one space

File: lib/logasm/preprocessors/whitelist.rb [hunk 8]
```diff
@@ -4,16 +4,24 @@ class Whitelist
...
+         .map(&method(:decode))
+        @fields_to_include = decoded_pointers.inject({}) do |mem, pointer|
+          mem.merge(pointer => true)
```
Comment (indrekj) [inline lib/logasm/preprocessors/whitelist.rb]: why do we need to make it a hash? It looks like it is always `=> true`. maybe a simple map would suffice?

Comment (indrekj) [inline lib/logasm/preprocessors/whitelist.rb]: in the next block instead of ` +            mem.merge(get_wildcard_roots_of(pointer))` you could do `mem + get_wildcard_roots_of(pointer)` assuming get_wildcard_roots_of can return an array

Comment (ostap0207) [inline lib/logasm/preprocessors/whitelist.rb]: What do you mean by

>simple map

Comment (indrekj) [inline lib/logasm/preprocessors/whitelist.rb]: something like
```ruby
@fields_to_include = decoded_pointers
@wildcards = decoded_pointers
  .select(&method(:has_wildcard?))
  .inject([]) do |mem, pointer|
    mem + get_wildcard_roots_of(pointer)
  end
```

and in `get_wildcard_roots_of` you return `[ponter]` or `[pointer1, pointer2, ..]`

Comment (indrekj) [inline lib/logasm/preprocessors/whitelist.rb]: but if you think there could be performance problems then this solution is also ok

Comment (ostap0207) [inline lib/logasm/preprocessors/whitelist.rb]: So it's gonna be an array in this case. I need hash to be able to find if the pointer exists on not, e.g. `@fields_to_include[parent_pointer]`, same with wildcards. I don't think array is good for this situation.

Comment (indrekj) [inline lib/logasm/preprocessors/whitelist.rb]: you could do `@fields_to_include.include?(parent_pointer)` but okay, hashmaps are faster.

**Comment by indrekj** [comment]
+1 after my comments are addressed

**Comment by ostap0207** [comment]
Updated. @ChameleonTartu added commit message, would leave the formatting though.

**Comment by indrekj** [comment]
!squash
